### ISSUE TITLE
Integrate sklearn benchmark information to refine TPOT classifier parameters

### ISSUE THREAD

**Issue by rhiever** [rhiever/tpot#161]
We have the results in from the sklearn benchmark, and we have some decent ideas on what parameters should be tuned for each model and in what ranges. Let's incorporate this information into TPOT to help it with the optimization process.

If I don't mention a particular parameter that's already evolvable for a classifier, then that means the parameter can be left as-is.

# AdaBoostClassifier

Fix `n_estimators` at 500; no longer evolvable

Force `learning_rate` in the range `[0.0001, 1.]`

# BernoulliNB

Fix `fit_prior` to `True`; no longer evolvable

# DecisionTreeClassifier

Remove `max_depth` and `max_features` as evolvable parameters; use the sklearn defaults

Add `min_weight_fraction_leaf` as an evolvable  parameter in the range `[0., 0.5]` (float)

# ExtraTreesClassifier

Change `max_features` to be a float in the range `[0., 1.]`

Remove `max_depth` and use the sklearn default

Add `min_weight_fraction_leaf` as an evolvable parameter in the range `[0., 0.5]` (float)

# GradientBoostingClassifier

Force `learning_rate` in the range `[0.0001, 1.]`

Add `max_features` as an evolvable parameter in the range `[0., 1.]` (float)

Add `min_weight_fraction_leaf` as an evolvable parameter in the range `[0., 0.5]` (float)

Add `loss` as an evolvable parameter w/ options `['exponential', 'deviance']`

Remove `max_depth` and use the sklearn default

# KNeighborsClassifier

Add `weights` as an evolvable parameter w/ options `['uniform', 'distance']`

# LinearSVC

Remove `fit_intercept` and `tol` from the evolvable parameters and use the sklearn defaults

Force `C` in the range `[0.0001, 25.]`

Add `penalty` as an evolvable parameter w/ options `['l1', 'l2']`

Add `dual` as an evolvable parameter w/ options `[True, False]`

# LogisticRegression

Force `C` in the range `[0.0001, 50.]`

Add `penalty` as an evolvable parameter w/ options `['l1', 'l2']`

Add `dual` as an evolvable parameter w/ options `[True, False]`

# MultinomialNB

Fix `fit_prior` to `True`; no longer evolvable

# PassiveAggressiveClassifier

Remove `fit_intercept` from the evolvable parameters and fix it to `True`

Force `C` in the range `[0.0001, 1.]`

# RandomForestClassifier

Remove `max_features` from the evolvable parameters and use the sklearn default

Add `min_weight_fraction_leaf` as an evolvable parameter in the range `[0., 0.5]` (float)

# SVC

Remove this classifier entirely from TPOT for now; too many parameters to simultaneously optimize

**Comment by rhiever** [comment]
cc @teaearlgraycold

**Comment by teaearlgraycold** [comment]
Any good alternatives to threshholding `int`s and `float`s besides just min/max-ing them?

I feel like taking the wild values that GP can provide us with and just doing `max(1.0, min(0., max_features))` won't be ideal. You'd just end up with a bunch of 1s and 0s.

What if I added a new terminal type, something like:

```Python
class float_limited(float):
    def __init__(self):
        pass

. . .

for val in [1.0, 0.1, 0.01, 0.001, 0.0001]:
            self._pset.addTerminal(val, float_limited)

. . .

self._pset.addPrimitive(self._extra_trees, [pd.DataFrame, int, float_limited], pd.DataFrame)
```

For everything that should be in a distribution like 1 to 50, we could simply multiply the `float_limited` by some scalar in the operator, so in that case `50`.

---

Alternatively we could try to make **all** floats in the 1.0 to 0.0001 range and then scale them appropriately as mentioned above.

**Comment by rhiever** [comment]
I've been thinking about removing the mathematical operators in favor of having a fixed terminal set. I think that in addition to being prone to producing crazy values, having so many mathematical operators in the TPOT trees causes most of the mutations and crossovers to happen on the mathematical operators rather than the pipeline operators, which is probably bad. (I still need to test that experimentally)

Let's just apply the caps for now and we'll raise a separate issue for dealing with the terminal set.

BTW: We should add booleans to the terminal set for the True/False parameters.

**Comment by teaearlgraycold** [comment]
>having so many mathematical operators in the TPOT trees causes most of the mutations and crossovers to happen on the mathematical operators rather than the pipeline operators, which is probably bad.

That's what I've seen.

>BTW: We should add booleans to the terminal set for the True/False parameters.

I agree

**Comment by rhiever** [comment]
Great. Let's add the boolean terminals in this PR.

**Comment by amueller** [comment]
for adaboost you might not want to fix n_estimators

**Comment by amueller** [comment]
you don't do it for gradient boosting, right?

**Comment by rhiever** [comment]
Yes, we fix `n_estimators=500` for the GradientBoostingClassifier because the docs claim:

>The number of boosting stages to perform. Gradient boosting is fairly robust to over-fitting so a large number usually results in better performance.

And I felt sage fixing `n_estimators=500` for AdaBoostClassifier because the docs say (emphasis mine):

>The **maximum** number of estimators at which boosting is terminated. In case of perfect fit, the learning procedure is stopped early.

**Comment by amueller** [comment]
huh alright then ^^

**Comment by rhiever** [comment]
Why do you disagree? :-) I'd say it was even supported by what we found from the benchmark.

**Comment by amueller** [comment]
Actually I agree ;)
